lib/gmp/gmp.nit

```diff
@@ -0,0 +1,370 @@
+        return new Long(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_d self
+        return new Ratio(tmp)
+    end
+end
+
+redef class Int
+    redef fun to_l do
+        var tmp = new NativeMPZ
+        tmp.set_si self
+        return new Long(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_si(self, 1)
+        return new Ratio(tmp)
+    end
+end
+
+class Long
+    super Discrete
+    super Numeric
+    super FinalizableOnce
+
+    redef type OTHER: Long
+
+    private var val: NativeMPZ
+
+    redef fun successor(l) do return self + l.to_l
+    redef fun predecessor(l) do return self - l.to_l
+
+    redef fun hash do return self.to_i
+
+    redef fun <=>(l) do
+        if (val.cmp(l.val)) < 0 then
+            return -1
+        else if (val.cmp(l.val)) > 0 then
+            return 1
+        else
+            return 0
+        end
+    end
+
+    redef fun ==(l) do
+        if l isa Long then
+            return (self <=> l) == 0
+        else
+            return false
+        end
+    end
+
+    redef fun <=(l) do return (self <=> l) <= 0
+    redef fun <(l) do return (self <=> l) < 0
+    redef fun >=(l) do return (self <=> l) >= 0
+    redef fun >(l) do return (self <=> l) > 0
+
+
+    #     assert(2.to_l + 2.to_l == 4.to_l)
+    redef fun +(l) do
+        var res = new NativeMPZ
+        val.add(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(-(2.to_l) == (-2).to_l)
+    redef fun - do
+        var res = new NativeMPZ
+        val.neg res
+        return new Long(res)
+    end
+
+    #     assert(2.to_l - 2.to_l == 0.to_l)
+    redef fun -(l) do
+        var res = new NativeMPZ
+        val.sub(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(2.to_l * 2.to_l == 4.to_l)
+    redef fun *(l) do
+        var res = new NativeMPZ
+        val.mul(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(3.to_l / 2.to_l == 1.to_l)
+    redef fun /(l) do
+        var res = new NativeMPZ
+        val.tdiv_q(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(5.to_l % 2.to_l == 1.to_l)
+    fun %(l: Long): Long do
+        var res = new NativeMPZ
+        val.mod(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(3.to_l ^ 2 == 9.to_l)
+    fun ^(i: Int): Long do
+        var res = new NativeMPZ
+        var pow = new UInt64
+        pow.set_si i
+        val.pow_ui(res, pow)
+        pow.free
+        return new Long(res)
+    end
+
+    #     assert((-3).to_l.abs == 3.to_l)
+    fun abs: Long do
+        var res = new NativeMPZ
+        val.abs res
+        return new Long(res)
+    end
+
+    #     assert(15.to_l.gcd(10.to_l) == 5.to_l)
+    fun gcd(l: Long): Long do
+        var res = new NativeMPZ
+        val.gcd(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert((0x10001).to_l.probab_prime(15) == 2)
+    fun probab_prime(reps: Int): Int do
```

Comment:
As this is high-level, I would like some doc, `probab_prime` does not mean too much to me at the moment